### PR TITLE
Create user login

### PR DESCRIPTION
**Pull request by jhubaum**
This PR contains two changes.
- Add a user login in the frontend. As a potential safe guard, add a config flag to disable logins. The safe guard can be set in the deployment config
- Add `gcadmin`. This is an executable to run basic admin tasks from the cli in the docker container. Right now, it supports adding and listing users, backing up the database, and triggering a data update

## Pre-merge checklist
- [ ] Squash fixup commits
- [x] Make sure that `gcadmin backup` works

## Review needs
- What do you think about optional email addresses for users?

### CODE REVIEW

File: src/users.py [hunk 1]
```diff
@@ -153,21 +178,66 @@ def register(self, username, password, *, session=None):
...
+    create_parser = subparsers.add_parser("create", help="Create new user")
+    create_parser.add_argument("username")
+    create_parser.add_argument("email", nargs="?")
+    create_parser.set_defaults(func=_create_new_user)
+
+
+def main():
+    parser = ArgumentParser("Control the task queue")
+    setup_parser(parser)
+    args = create_argparser().parse_args()
```
Comment (jhubaum) [inline src/users.py:238]: `create_argparser` no longer exists, which means this is broken (same for tasks.py).

**Comment by jhubaum** [comment]
@philippschmalen This is ready to be merged (after I squashed the commits). Do you want to take a look? If not, I will merge it some time this weekend.

**Comment by philippschmalen** [comment]
Again, amazing work! It's far from my competencies to say something about improvements, so I just provide a user perspective.
Docker container does not seem to communicate via HTTP with the plug. Tried to connect and run test mode, which did not react.

Maybe you can give me some tutorial on the server side, where I could run `gcadmin`.

Questions
* What does the config flag to disable logins prevent concretely?
* Why should email be optional? I see it as a primary identifier for users. IMO email=username and a "verify email" is a robust approach to create users and verify that they exist. Though not sure on that one, happy to hear your thoughts!

**Comment by jhubaum** [comment]
Yeah, I'll fix the docker communication in another MR (where I'll also switch to a self-hosted MQTT server). And I'm happy to walk you through all the server side stuff!

About feedback in general: I'm aware that concrete coding improvements are a hard ask. I'm more interested in exactly the user-perspective focused questions you're asking. Also, if there's some code you don't understand, or don't understand a particular choice I made, it's useful to bring it up here. As for your questions:
> What does the config flag to disable logins prevent concretely?

When we talked about user management the last time, one of us (I don't remember who exactly) was concerned that too many people could create an account and overwhelm the server (which would be a very nice problem to have :D). Whether or not these concerns are justified (they are probably not), this flag is a simple solution to this that was easy to implement. But it's probably not necessary. If you want, I can remove it again.

> Why should email be optional? 

I personally like it if I can try an app without needing to give my email first. Also, most features that require an email address (like resetting your password) are out of scope for now, because I don't want to figure out how to send automated emails in a safe way (i.e. in a way that doesn't risk blacklisting the entire server so that every mail lands in spam folders). Which means from the user perspective, I don't see why setting a mail is necessary. But I agree with your reasoning. Do you want me to make it non-optional?

**Comment by jhubaum** [comment]
@philippschmalen I added a fix for the test mode. Unfortunately hosting our own MQTT server is much more complicated than I anticipated, so we're stuck with a public MQTT broker. But at least the test mode is working now...

I'll add a fix for https://app.clickup.com/t/865bjfdqr to this PR and am going to merge afterwards. If you want me to require the email address for login, please say so today.

**Comment by jhubaum** [comment]
That's done. I'll leave this PR open until tomorrow to give you time to reply. Otherwise we can always create a follow-up.

**Comment by philippschmalen** [comment]
> Also, if there's some code you don't understand, or don't understand a particular choice I made, it's useful to bring it up here.

On design choices: Huge topic, and I probably lack fundamental understanding of it. That's where most of my question originate from, not being able to see through common patterns. Any resource you recommend to "catch up" a little?

> I personally like it if I can try an app without needing to give my email first

Totally agree and I like the simplicity, also as a user. Until we need security features, such as verify email, reset password, 2fa, we keep it that way.

> hosting our own MQTT server is much more complicated

Yes, that's a bottleneck. I feel a managed mqtt broker gives us peace of mind. Why not go with the [free tier of hivemq](https://www.hivemq.com/mqtt-cloud-broker/#:~:text=card%20information%20required.-,FREE,-PAY%20AS%20YOU)

**Comment by jhubaum** [comment]
> On design choices: Huge topic, and I probably lack fundamental understanding of it. That's where most of my question originate from, not being able to see through common patterns. Any resource you recommend to "catch up" a little?

FWIW, and since I wrote most of the code I'm changing here, it's normal to struggle for feedback in this case. I don't know any resources that are immediately helpful, but I'll think about it.

> Why not go with the [free tier of hivemq](https://www.hivemq.com/mqtt-cloud-broker/#:~:text=card%20information%20required.-,FREE,-PAY%20AS%20YOU)

Excellent idea, I will look into this. Since this is unrelated to the changes in this PR, I'll do that in a follow-up.

**Comment by philippschmalen** [comment]
> Why not go with the [free tier of hivemq](https://www.hivemq.com/mqtt-cloud-broker/#:~:text=card%20information%20required.-,FREE,-PAY%20AS%20YOU)

I'm trying to connect the broker to the plug and test the plug // broker communication. You could focus on how to server // broker communications. For example, how to run [mqtt cli](https://hivemq.github.io/mqtt-cli/) commands on our server.

**Pull request merged**